drivers/gpio/gpio_b91.c

```diff
@@ -0,0 +1,564 @@
+	uint8_t analog_reg;
+
+	pin = BIT(pin);
+	val = up_down_res & 0x03;
+	analog_reg = 0x0e + (GET_PORT_NUM(gpio) << 1) + ((pin & 0xf0) ? 1 : 0);
+
+	if (pin & 0x11) {
+		val = val << 0;
+		mask = 0xfc;
+	} else if (pin & 0x22) {
+		val = val << 2;
+		mask = 0xf3;
+	} else if (pin & 0x44) {
+		val = val << 4;
+		mask = 0xcf;
+	} else if (pin & 0x88) {
+		val = val << 6;
+		mask = 0x3f;
+	} else {
+		return;
+	}
+
+	analog_write_reg8(analog_reg, (analog_read_reg8(analog_reg) & mask) | val);
+}
+
+/* Config Pin pull-up / pull-down resistors */
+static void gpio_b91_config_up_down_res(volatile struct gpio_b91_t *gpio,
+					gpio_pin_t pin,
+					gpio_flags_t flags)
+{
+	if ((flags & GPIO_PULL_UP) != 0) {
+		gpio_b91_up_down_res_set(gpio, pin, GPIO_PIN_PULLUP_10K);
+	} else if ((flags & GPIO_PULL_DOWN) != 0) {
+		gpio_b91_up_down_res_set(gpio, pin, GPIO_PIN_PULLDOWN_100K);
+	} else {
+		gpio_b91_up_down_res_set(gpio, pin, GPIO_PIN_UP_DOWN_FLOAT);
+	}
+}
+
+/* Config Pin In/Out direction */
+static void gpio_b91_config_in_out(volatile struct gpio_b91_t *gpio,
+				   gpio_pin_t pin,
+				   gpio_flags_t flags)
+{
+	uint8_t ie_addr = 0;
+
+	/* Port C and D Input Enable registers are located in another place: analog */
+	if (IS_PORT_C(gpio)) {
+		ie_addr = areg_gpio_pc_ie;
+	} else if (IS_PORT_D(gpio)) {
+		ie_addr = areg_gpio_pd_ie;
+	}
+
+	/* Enable/disable output */
+	WRITE_BIT(gpio->oen, pin, ~flags & GPIO_OUTPUT);
+
+	/* Enable/disable input */
+	if (ie_addr != 0) {
+		/* Port C and D are located in analog space */
+		if (flags & GPIO_INPUT) {
+			analog_write_reg8(ie_addr, analog_read_reg8(ie_addr) | BIT(pin));
+		} else {
+			analog_write_reg8(ie_addr, analog_read_reg8(ie_addr) & (~BIT(pin)));
+		}
+	} else {
+		/* Input Enable registers of all other ports are located in common GPIO space */
+		WRITE_BIT(gpio->ie, pin, flags & GPIO_INPUT);
+	}
+}
+
+/* GPIO driver initialization */
+static int gpio_b91_init(const struct device *dev)
+{
+	const struct gpio_b91_config *cfg = GET_CFG(dev);
+
+	cfg->pirq_connect();
+
+	return 0;
+}
+
+/* API implementation: pin_configure */
+static int gpio_b91_pin_configure(const struct device *dev,
+				  gpio_pin_t pin,
+				  gpio_flags_t flags)
+{
+	volatile struct gpio_b91_t *gpio = GET_GPIO(dev);
+
+	/* Check input parameters: pin number */
+	if (pin > PIN_NUM_MAX) {
+		return -ENOTSUP;
+	}
+
+	/* Check input parameters: open-source and open-drain */
+	if ((flags & GPIO_SINGLE_ENDED) != 0) {
+		return -ENOTSUP;
+	}
+
+	/* Check input parameters: simultaneous in/out mode */
+	if ((flags & GPIO_OUTPUT) && (flags & GPIO_INPUT)) {
+		return -ENOTSUP;
+	}
+
+	/* Set GPIO init state if defined to avoid glitches */
+	if ((flags & GPIO_OUTPUT_INIT_HIGH) != 0) {
+		gpio->output |= BIT(pin);
+	} else if ((flags & GPIO_OUTPUT_INIT_LOW) != 0) {
+		gpio->output &= ~BIT(pin);
+	}
+
+	/* GPIO function enable */
+	WRITE_BIT(gpio->actas_gpio, BIT(pin), 1);
+
+	/* Setting Drive Strength is not supported by Zephyr */
```

Comment:
Actually, they are supported:
https://github.com/zephyrproject-rtos/zephyr/blob/be2e6a08500ef8c157639b89d27c0a858ec43dcb/include/drivers/gpio.h#L212

Reply:
Fixed. 
Removed this comment and added verification for drive strength. Currently we are not supporting it